File: app/src/main/java/com/nicholasworkshop/tinklabstest/fragment/guide/GuideView.java
package com.nicholasworkshop.tinklabstest.fragment.guide;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.pwittchen.infinitescroll.library.InfiniteScrollListener;
import com.nicholasworkshop.tinklabstest.R;
import com.nicholasworkshop.tinklabstest.external.ads.model.Ad;
import com.nicholasworkshop.tinklabstest.external.content.model.Story;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

/**
 * Created by nicholas.wong on 2017/05/15.
 */
public class GuideView {

    private final GuideRecyclerViewAdapter guideRecyclerViewAdapter;

    @BindView(R.id.stories) RecyclerView storiesRecyclerView;

    @Inject
    GuideView(GuideRecyclerViewAdapter guideRecyclerViewAdapter) {
        this.guideRecyclerViewAdapter = guideRecyclerViewAdapter;
    }

    View createView(LayoutInflater inflater, ViewGroup container) {
        View view = inflater.inflate(R.layout.fragment_guide, container, false);
        ButterKnife.bind(this, view);
        LinearLayoutManager layoutManager = new LinearLayoutManager(container.getContext());
        storiesRecyclerView.setHasFixedSize(true);
        storiesRecyclerView.setLayoutManager(layoutManager);
        storiesRecyclerView.setAdapter(guideRecyclerViewAdapter);
        storiesRecyclerView.addOnScrollListener(new GuideInfiniteScrollListener(1, layoutManager));
        return view;
    }

    void setStoryList(List<Story> storyList) {
        guideRecyclerViewAdapter.setStoryList(storyList);
    }

    void setAdList(List<Ad> adList) {
        guideRecyclerViewAdapter.setAdList(adList);
    }

    private Subject<Integer> storyRequestSubject = BehaviorSubject.create();

    Observable<Integer> storyRequests() {
        return storyRequestSubject;
    }

    private class GuideInfiniteScrollListener extends InfiniteScrollListener {

        GuideInfiniteScrollListener(int maxItemsPerRequest, LinearLayoutManager layoutManager) {
            super(maxItemsPerRequest, layoutManager);
        }

        @Override
        public void onScrolledToEnd(int firstVisibleItemPosition) {
            Log.d("dsafsdaf", "fdsafsadf" + firstVisibleItemPosition);
            storyRequestSubject.onNext(firstVisibleItemPosition);
        }
    }
}

File: app/src/main/kotlin/com/nicholasworkshop/tinklabstest/fragment/guide/GuideView.kt
package com.nicholasworkshop.tinklabstest.fragment.guide

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.pwittchen.infinitescroll.library.InfiniteScrollListener
import com.nicholasworkshop.tinklabstest.R
import com.nicholasworkshop.tinklabstest.external.ads.model.Ad
import com.nicholasworkshop.tinklabstest.external.content.model.Story
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.fragment_guide.*
import javax.inject.Inject

class GuideView @Inject internal constructor(
        private val guideRecyclerViewAdapter: GuideRecyclerViewAdapter)
    : LayoutContainer {

    override lateinit var containerView: View

    lateinit var storiesView: RecyclerView
    lateinit var layoutManager: LinearLayoutManager

    private val storyRequestSubject = BehaviorSubject.create<Int>()

    internal fun createView(inflater: LayoutInflater, container: ViewGroup): View {
        containerView = inflater.inflate(R.layout.fragment_guide, container, false)
        layoutManager = LinearLayoutManager(containerView.context)
        storiesView = stories_view
        return containerView
    }

    internal fun viewCreated() {
        storiesView.setHasFixedSize(true)
        storiesView.layoutManager = layoutManager
        storiesView.adapter = guideRecyclerViewAdapter
        storiesView.addOnScrollListener(GuideInfiniteScrollListener(1, layoutManager))
    }

    internal fun setStoryList(storyList: List<Story>) {
        guideRecyclerViewAdapter.setStoryList(storyList)
    }

    internal fun setAdList(adList: List<Ad>) {
        guideRecyclerViewAdapter.setAdList(adList)
    }

    internal fun storyRequests(): Observable<Int> {
        return storyRequestSubject
    }

    internal inner class GuideInfiniteScrollListener(maxItemsPerRequest: Int, layoutManager: LinearLayoutManager)
        : InfiniteScrollListener(maxItemsPerRequest, layoutManager) {

        override fun onScrolledToEnd(firstVisibleItemPosition: Int) {
            storyRequestSubject.onNext(layoutManager.findLastVisibleItemPosition())
        }
    }
}
